Image ros packages patches

ivanpauno opened the pull request:
This adds some ros image related packages ->
- image_common
  * image_common
  * camera_info_manager
  * camera_calibration_parsers
  * image_transport
  * polled_camera

- image_pipeline
  * image_view
  * camera_calibration
  * image_pipeline
  * depth_image_proc
  * image_proc
  * image_publisher
  * image_rotate
  * stereo_image_proc

- image_transport_plugins
  * compressed_depth_image_transport
  * compressed_image_transport
  * image_transport_plugins

- vision_opencv
  * image_geometry
  * vision_opencv

ivanpauno committed:
Added some ros image related packages

ivanpauno committed:
Uncommented image-transport patch

ivanpauno committed:
Patched image view

meyerj commented:
Please hold back, I am going to propose some additional patches.

Edit: see https://github.com/Intermodalics/ros_android/pull/19. I am not sure whether it fixes all remaining issues. It only targets `camera_calibration_parsers`.

meyerj committed:
Modified camera_calibration_parsers patch to fix static linking issues

ivanpauno committed:
Merge pull request #19 from Intermodalics/meyerj/image-ros-packages-patches

Modified camera_calibration_parsers patch to fix static linking issues

ivanpauno committed:
Added image_proc, depth_image_proc and stereo_image_proc. Solved transitive linking problem with bondcpp.

ivanpauno committed:
Added image_publisher. Solved linking problems.

ivanpauno committed:
Added image rotate package

jubeira commented on ndk.rosinstall:
```diff
@@ -682,15 +682,15 @@
+    uri: https://github.com/ros-gbp/vision_opencv-release/archive/release/kinetic/vision_opencv/1.12.8-0.tar.gz
+    version: vision_opencv-release-release-kinetic-vision_opencv-1.12.8-0
+- tar:
+    local-name: xacro
```
xacro fails saying it doesn't find `roslint`. Please uncomment `roslint` as well from this list.

ivanpauno replied:
I will comment again Xacro here, because I uncommented it by error.
Also, xacro works well if you uncomment the other dependencies needed.

meyerj replied:
`xacro` is Python-based. It might work because all Python packages use the Python interpreter installed on the host ([here](https://github.com/Intermodalics/ros_android/blob/cb2465a011ee25956fb463ddafdeafae11b42e5d/build_cpp.sh#L88-L91) and [here](https://github.com/Intermodalics/ros_android/blob/cb2465a011ee25956fb463ddafdeafae11b42e5d/utils.sh#L50)). On Android it is useless.

jubeira replied:
If there's no package depending on `xacro` (no package should depend on it at build time), please comment it out and add a comment saying it's python based.

jubeira commented on ndk.rosinstall:
```diff
@@ -222,70 +222,70 @@
+    uri: https://github.com/ros-gbp/image_pipeline-release/archive/release/kinetic/camera_calibration/1.12.23-0.tar.gz
+    version: image_pipeline-release-release-kinetic-camera_calibration-1.12.23-0
+- tar:
+    local-name: image_pipeline/depth_image_proc
```
This one fails asking for `eigen_conversions`; please uncomment it as well from this list.

ivanpauno replied:
I forgot uncommenting that, I will check again.

jubeira commented on ndk.rosinstall:
```diff
@@ -222,70 +222,70 @@
+    uri: https://github.com/ros-gbp/image_pipeline-release/archive/release/kinetic/image_rotate/1.12.23-0.tar.gz
+    version: image_pipeline-release-release-kinetic-image_rotate-1.12.23-0
+- tar:
+    local-name: image_pipeline/image_view
```
There's an issue with `camera_calibration_parsers` when linking here:

```
CMakeFiles/image_saver.dir/src/nodes/image_saver.cpp.o: In function `Callbacks::callbackWithCameraInfo(boost::shared_ptr<sensor_msgs::Image_<std::__ndk1::allocator<void> > const> const&, boost::shared_ptr<sensor_msgs::CameraInfo_<std::__ndk1::allocator<void> > const> const&)':
/opt/ros_android/output/catkin_ws/src/image_pipeline/image_view/src/nodes/image_saver.cpp:138: undefined reference to `camera_calibration_parsers::writeCalibration(std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&, std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&, sensor_msgs::CameraInfo_<std::__ndk1::allocator<void> > const&)'
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [/opt/ros_android/output/catkin_ws/devel/image_view/lib/image_view/image_saver] Error 1
```

The patch is being applied correctly though. Does this happen to you as well?

jubeira commented on ndk.rosinstall:
```diff
@@ -222,70 +222,70 @@
+    uri: https://github.com/ros-gbp/image_pipeline-release/archive/release/kinetic/image_proc/1.12.23-0.tar.gz
+    version: image_pipeline-release-release-kinetic-image_proc-1.12.23-0
+- tar:
+    local-name: image_pipeline/image_publisher
```
I'm getting an undefined reference when linking here:
```
/opt/ros_android/output/target/lib/libcamera_info_manager.a(camera_info_manager.cpp.o): In function `camera_info_manager::CameraInfoManager::loadCalibrationFile(std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&, std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&)':
/opt/ros_android/output/catkin_ws/src/image_common/camera_info_manager/src/camera_info_manager.cpp:277: undefined reference to `camera_calibration_parsers::readCalibration(std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&, std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >&, sensor_msgs::CameraInfo_<std::__ndk1::allocator<void> >&)'
/opt/ros_android/output/target/lib/libcamera_info_manager.a(camera_info_manager.cpp.o): In function `camera_info_manager::CameraInfoManager::saveCalibrationFile(sensor_msgs::CameraInfo_<std::__ndk1::allocator<void> > const&, std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&, std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&)':
/opt/ros_android/output/catkin_ws/src/image_common/camera_info_manager/src/camera_info_manager.cpp:546: undefined reference to `camera_calibration_parsers::writeCalibration(std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&, std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&, sensor_msgs::CameraInfo_<std::__ndk1::allocator<void> > const&)'
```
The patch seems to be applied correctly; does the same happen to you?

jubeira reviewed (changes requested):
Thanks @ivanpauno; this looks promising.

I'm getting a couple of errors; could you please take a look?

meyerj committed:
Fixed linking to camera_calibration_parsers

meyerj commented:
I am sorry. I accidentally removed the `LIBRARIES ${PROJECT_NAME}` line in #19. The patch should have added `DEPENDS ...` only and moved the `caktin_package()` call down to after finding dependencies.

Fixed in https://github.com/Intermodalics/ros_android/pull/16/commits/82781a4a80bd39d6db298d9520a8a755267c7f5b. Not tested yet.

It's a bit confusing to resolve merge conflicts in patch files ;-)

meyerj commented:
`image_publisher` is built as a shared library:
```
-rw-r--r-- 1 root root  113592464 Nov 27 03:52 ./libimage_publisher.so
```

ivanpauno committed:
Modified ndk.rosinstall

ivanpauno commented:
> I am sorry. I accidentally removed the `LIBRARIES ${PROJECT_NAME}` line in #19. The patch should have added `DEPENDS ...` only and moved the `caktin_package()` call down to after finding dependencies.
> 
> Fixed in [82781a4](https://github.com/Intermodalics/ros_android/commit/82781a4a80bd39d6db298d9520a8a755267c7f5b). Not tested yet.
> 
> It's a bit confusing to resolve merge conflicts in patch files ;-)

Thanks! I tested it and it's working fine.

ivanpauno committed:
Changed image_publisher to static library building. Also solved some transitive linking problems

ivanpauno committed:
Correcter path in patch files.

ivanpauno committed:
Solved merge conflicts with kinetic-wip

ivanpauno commented:
All commented problems have been solved.
After  82781a4a80bd39d6db298d9520a8a755267c7f5b linking problems commented by @jubeira are solved.
After fb712bb9525fd94fd93b761fbf9bf09310016945 image_publisher is built as a static library. I also corrected some CMakeList.txt to avoid future transitive linking problems.
I have also added eigen_conversions to ndk.rosinstall and commented out xacro.

Please @jubeira check if it is working for you. I've tested it without having problems.

jubeira reviewed (approved):
Tested; build passes!

I also didn't see any weird `so` around, and I checked the output libraries and the architecture was fine. Good job!
